### PR TITLE
Mvi id status

### PR DESCRIPTION
**Pull request by annaswims**
## Description of change
based upon my [understanding of the MVI documentation](https://github.com/department-of-veterans-affairs/vets.gov-team/issues/18388) we should on use id's with a status of A or P.

## Testing done
wip

## Acceptance Criteria (Definition of Done)
merge aftter https://github.com/department-of-veterans-affairs/vets-api/pull/3219

#### Applies to all PRs

- [x] Appropriate logging
- [x] Swagger docs have been updated, if applicable
- [x] Provide link to originating GitHub issue, or connected to it via ZenHub
- [x] Does not contain any sensitive information (i.e. PII/credentials/internal URLs/etc., in logging, hardcoded, or in specs)
- [x] Provide which alerts would indicate a problem with this functionality (if applicable)

### CODE REVIEW

File: lib/mvi/responses/id_parser.rb [hunk 1]
```diff
@@ -12,28 +12,35 @@ class IdParser
...
+      #              D = Deprecated from a Duplicate
+      #              M = Deprecated from a Mismatch
+      #              U - Deprecated from an Unlink
+      #              L = pending local merge - This is a value identified for correlations including EDIPI when the
+      #                   identifier has been deprecated/associated to another active ID.
+      #              PCE = Pending Cat Edit correlations (unsure if this should be used, likely not)
 
       def parse(ids)
         ids = ids.map(&:attributes)
-
```
Comment (annaswims) [inline lib/mvi/responses/id_parser.rb]: probably want to [ignore whitespace](https://github.com/department-of-veterans-affairs/vets-api/pull/3259/files?w=1) when viewing the diff

File: lib/mvi/responses/id_parser.rb [hunk 2]
```diff
@@ -5,15 +5,15 @@ module Responses
...
+      ICN_REGEX = /^\w+\^NI\^\w+\^\w+\^\w+$/
       VET360_ASSIGNING_AUTHORITY_ID = '^NI^200M^USVHA'
 
       # MVI correlation id source id relationships:
       # {source id}^{id type}^{assigning facility}^{assigning authority}^{id status}
       # id type - NI = national identifier
       #           PI = patient identifier
 
-      # id status -  A = active
+      # id status -  A = Active
```
Comment (LindseySaari) [inline lib/mvi/responses/id_parser.rb]: ➕1 For consistency 🎉

File: lib/mvi/responses/id_parser.rb [hunk 3]
```diff
@@ -12,28 +12,35 @@ class IdParser
...
-          active_mhv_ids: select_ids(select_extension(ids, /^\w+\^PI\^200MH.{0,1}\^\w+\^A$/, CORRELATION_ROOT_ID)),
-          edipi: select_ids(select_extension(ids, /^\w+\^NI\^200DOD\^USDOD\^\w+$/, EDIPI_ROOT_ID))&.first,
-          vba_corp_id: select_ids_except(select_extension(ids, /^\w+\^PI\^200CORP\^USVBA\^\w+$/, CORRELATION_ROOT_ID),
-                                         %w[H L])&.first,
+          icn:              select_ids(select_extension(ids, ICN_REGEX,                           CORRELATION_ROOT_ID))&.first,
+          sec_id:           select_ids(select_extension(ids, /^\w+\^PN\^200PROV\^USDVA\^\w+$/,    CORRELATION_ROOT_ID))&.first,
+          mhv_ids:          select_ids(select_extension(ids, /^\w+\^PI\^200MH.{0,1}\^\w+\^\w+$/,  CORRELATION_ROOT_ID)),
+          active_mhv_ids:   select_ids(select_extension(ids, /^\w+\^PI\^200MH.{0,1}\^\w+\^A$/,    CORRELATION_ROOT_ID)),
+          edipi:            select_ids(select_extension(ids, /^\w+\^NI\^200DOD\^USDOD\^\w+$/,     EDIPI_ROOT_ID))&.first,
+          vba_corp_id:      select_ids(select_extension(ids, /^\w+\^PI\^200CORP\^USVBA\^(A|P)$/,  CORRELATION_ROOT_ID))&.first,
```
Comment (johnpaulashenfelter) [inline lib/mvi/responses/id_parser.rb]: The `A or P` statuses are now checked inline here with that trailing regex. Seems consistent enough with the rest of this, though they are all a bit complicated to parse mentally (at least for me).

I'd love to see @jholton take a quick look at this since he did the work in the corresponding PR you mention and likely has the most context

👍 for deleting more code!

Comment (jholton) [reply lib/mvi/responses/id_parser.rb]: I'm on it.  I haven't reviewed the PR yet, because I'm currently trying to get more information about these IDs (and statuses) from MVI engineers.

File: lib/mvi/responses/id_parser.rb [hunk 4]
```diff
@@ -5,35 +5,42 @@ module Responses
     class IdParser
       CORRELATION_ROOT_ID = '2.16.840.1.113883.4.349'
       EDIPI_ROOT_ID = '2.16.840.1.113883.3.42.10001.100001.12'
-      ICN_REGEX = /^\w+\^NI\^\w+\^\w+\^\w+$/.freeze
+      ICN_REGEX = /^\w+\^NI\^200M\^USVHA\^P$/.freeze
```
Comment (jholton) [inline lib/mvi/responses/id_parser.rb]: This was matching any ID that had an ID Type of 'NI'.  
In every response I've seen from MVI, the first ID is the ICN so this might have never been a practical problem.

From communication with an MVI engineer:
> Will all ICNs have an ID Type of 'NI', assigning authority of '200M' and assigning facility of 'USVHA'?
> > Yes

File: lib/mvi/responses/id_parser.rb [hunk 5]
```diff
@@ -5,35 +5,42 @@ module Responses
...
       # id type - NI = national identifier
       #           PI = patient identifier
-      # id status -  A = active
-      #              P = permanent
-      #              H = deprecated(do not use)
-      #              L = pending local merge (do not use)
-      #              PCE = ??
+
+      # id status -  A = Active (only applies to correlation IDs)
+      #              P = Permanent (only applies to ICNs)
```
Comment (jholton) [inline lib/mvi/responses/id_parser.rb]: For clarification:
From my research you can divide the IDs into one of two categories:
* ICN
* Correlation ID

An ICN is the MVI internal ID.
All other IDs are Correlation IDs.  That is, they are IDs that correlate with some other system.

An ICN can have an ID Status of 'P', 'T' or 'D' but we will only see ICNs with a status of 'P' in the IDs returned at the `//controlActProcess/subject/registrationEvent/subject1/patient/id` path in the response XML.  Those are the IDs that are parsed here.
In short, every successful response will have an ICN that matches `/^\w+\^NI\^200M\^USVHA\^P$/`

The other IDs are Correlation IDs.
These IDs can have various ID Types, Assigning Authorities and Assigning Facility.
They can have various status IDs as mentioned below: A, D, M, U, L, PCE, etc.
Only 'A' designates a valid, current ID.

File: lib/mvi/responses/id_parser.rb [hunk 6]
```diff
@@ -3,36 +3,44 @@
...
     class IdParser
-      CORRELATION_ROOT_ID = '2.16.840.1.113883.4.349'
-      EDIPI_ROOT_ID = '2.16.840.1.113883.3.42.10001.100001.12'
-      ICN_REGEX = /^\w+\^NI\^\w+\^\w+\^\w+$/.freeze
-      VET360_ASSIGNING_AUTHORITY_ID = '^NI^200M^USVHA'
+      VA_ROOT_OID = '2.16.840.1.113883.4.349'
+      DOD_ROOT_OID = '2.16.840.1.113883.3.42.10001.100001.12'
+      ICN_REGEX = /^\w+\^NI\^200M\^USVHA\^\w+$/.freeze
+      PERMANENT_ICN_REGEX = /^\w+\^NI\^200M\^USVHA\^P$/.freeze
+      ICN_ASSIGNING_AUTHORITY_ID = '^NI^200M^USVHA'
```
Comment (jholton) [inline lib/mvi/responses/id_parser.rb]: Renaming constants to more accurately reflect what they are in MVI terms

File: lib/mvi/responses/icn_with_aaid_parser.rb [hunk 7]
```diff
@@ -28,17 +27,13 @@ def initialize(extension)
       #
       def without_id_status
         return if extension.nil?
-        return if invalid_id_status?
+        return unless id_status == 'P'
```
Comment (jholton) [inline lib/mvi/responses/icn_with_aaid_parser.rb]: Should I leave this line?
This class is only used in `MVI::Responses::IdParser`, which now only passes in IDs with a status of 'P'

File: lib/mvi/responses/id_parser.rb [hunk 8]
```diff
@@ -3,36 +3,44 @@
...
-          mhv_ids: select_ids(select_extension(ids, /^\w+\^PI\^200MH.{0,1}\^\w+\^\w+$/, CORRELATION_ROOT_ID)),
-          active_mhv_ids: select_ids(select_extension(ids, /^\w+\^PI\^200MH.{0,1}\^\w+\^A$/, CORRELATION_ROOT_ID)),
-          edipi: select_ids(select_extension(ids, /^\w+\^NI\^200DOD\^USDOD\^\w+$/, EDIPI_ROOT_ID))&.first,
-          vba_corp_id: select_ids_except(select_extension(ids, /^\w+\^PI\^200CORP\^USVBA\^\w+$/, CORRELATION_ROOT_ID),
-                                         %w[H L])&.first,
-          vha_facility_ids: select_facilities(select_extension(ids, /^\w+\^PI\^\w+\^USVHA\^\w+$/, CORRELATION_ROOT_ID)),
-          birls_id: select_ids(select_extension(ids, /^\w+\^PI\^200BRLS\^USVBA\^\w+$/, CORRELATION_ROOT_ID))&.first,
-          vet360_id: select_ids(select_extension(ids, /^\w+\^PI\^200VETS\^USDVA\^A$/, CORRELATION_ROOT_ID))&.first,
-          icn_with_aaid: ICNWithAAIDParser.new(full_icn_with_aaid(ids)).without_id_status
+          icn:              select_ids(select_extension(ids, PERMANENT_ICN_REGEX,                 VA_ROOT_OID))&.first,
```
Comment (jholton) [inline lib/mvi/responses/id_parser.rb]: Changed regex to require 'P' ID status

File: lib/mvi/responses/id_parser.rb [hunk 9]
```diff
@@ -3,36 +3,44 @@
...
+          sec_id:           select_ids(select_extension(ids, /^\w+\^PN\^200PROV\^USDVA\^A$/,      VA_ROOT_OID))&.first,
```
Comment (jholton) [inline lib/mvi/responses/id_parser.rb]: Changed regex to require 'A' ID status

File: lib/mvi/responses/id_parser.rb [hunk 10]
```diff
@@ -3,36 +3,44 @@
...
+          mhv_ids:          select_ids(select_extension(ids, /^\w+\^PI\^200MH.{0,1}\^\w+\^\w+$/,  VA_ROOT_OID)),
+          active_mhv_ids:   select_ids(select_extension(ids, /^\w+\^PI\^200MH.{0,1}\^\w+\^A$/,    VA_ROOT_OID)),
```
Comment (jholton) [inline lib/mvi/responses/id_parser.rb]: No change to regex since these cover Active and inactive ID statuses already.

File: lib/mvi/responses/id_parser.rb [hunk 11]
```diff
@@ -3,36 +3,44 @@
...
+          edipi:            select_ids(select_extension(ids, /^\w+\^NI\^200DOD\^USDOD\^A$/,       DOD_ROOT_OID))&.first,
```
Comment (jholton) [inline lib/mvi/responses/id_parser.rb]: Changed regex to require 'A' ID status

File: lib/mvi/responses/id_parser.rb [hunk 12]
```diff
@@ -3,36 +3,44 @@
...
+          vba_corp_id:      select_ids(select_extension(ids, /^\w+\^PI\^200CORP\^USVBA\^A$/,      VA_ROOT_OID))&.first,
```
Comment (jholton) [inline lib/mvi/responses/id_parser.rb]: Changed regex to require 'A' ID status

File: lib/mvi/responses/id_parser.rb [hunk 13]
```diff
@@ -3,36 +3,44 @@
...
+          vha_facility_ids: select_facilities(select_extension(ids, /^\w+\^PI\^\w+\^USVHA\^\w+$/, VA_ROOT_OID)),
+          birls_id:         select_ids(select_extension(ids, /^\w+\^PI\^200BRLS\^USVBA\^A$/,      VA_ROOT_OID))&.first,
```
Comment (jholton) [inline lib/mvi/responses/id_parser.rb]: Changed regex to require 'A' ID status

File: lib/mvi/responses/id_parser.rb [hunk 14]
```diff
@@ -3,36 +3,44 @@
...
+          vet360_id:        select_ids(select_extension(ids, /^\w+\^PI\^200VETS\^USDVA\^A$/,      VA_ROOT_OID))&.first,
```
Comment (jholton) [inline lib/mvi/responses/id_parser.rb]: Changed regex to require 'A' ID status

File: lib/mvi/responses/id_parser.rb [hunk 15]
```diff
@@ -3,36 +3,44 @@
...
-          birls_id: select_ids(select_extension(ids, /^\w+\^PI\^200BRLS\^USVBA\^\w+$/, CORRELATION_ROOT_ID))&.first,
-          vet360_id: select_ids(select_extension(ids, /^\w+\^PI\^200VETS\^USDVA\^A$/, CORRELATION_ROOT_ID))&.first,
-          icn_with_aaid: ICNWithAAIDParser.new(full_icn_with_aaid(ids)).without_id_status
+          icn:              select_ids(select_extension(ids, PERMANENT_ICN_REGEX,                 VA_ROOT_OID))&.first,
+          sec_id:           select_ids(select_extension(ids, /^\w+\^PN\^200PROV\^USDVA\^A$/,      VA_ROOT_OID))&.first,
+          mhv_ids:          select_ids(select_extension(ids, /^\w+\^PI\^200MH.{0,1}\^\w+\^\w+$/,  VA_ROOT_OID)),
+          active_mhv_ids:   select_ids(select_extension(ids, /^\w+\^PI\^200MH.{0,1}\^\w+\^A$/,    VA_ROOT_OID)),
+          edipi:            select_ids(select_extension(ids, /^\w+\^NI\^200DOD\^USDOD\^A$/,       DOD_ROOT_OID))&.first,
+          vba_corp_id:      select_ids(select_extension(ids, /^\w+\^PI\^200CORP\^USVBA\^A$/,      VA_ROOT_OID))&.first,
+          vha_facility_ids: select_facilities(select_extension(ids, /^\w+\^PI\^\w+\^USVHA\^\w+$/, VA_ROOT_OID)),
```
Comment (jholton) [inline lib/mvi/responses/id_parser.rb]: Did _not_ change regex to require 'A'. 
Presence of IDs is checked [here](https://github.com/department-of-veterans-affairs/vets-api/blob/91b76792905302a5eb5eb2b4d7661050af54dd89/app/models/user.rb#L182) to determine if a user is a VA patient. 
The individual IDs themselves are not used anywhere AFAICT.

**Comment by annaswims** [comment]
The code looks good to me, but I'd like to talk a bit about how we can validate the changes have the intended effect.  Can we dig up some metrics that might help us validate that these changes have positive effects for users?  

The triage team (and perhaps the call center?) should be aware that we've made a change that might impact users in ways we haven't intended.

**Comment by annaswims** [comment]
@jholton Github won't allow me to approve this PR since I initiated it. But I do approve as soon as rubocop passes. I think that means you could approve.
